cmd/bad-key-revoker/main.go

```diff
@@ -461,29 +472,74 @@ func main() {
-		dbMap:           dbMap,
-		maxRevocations:  config.BadKeyRevoker.MaximumRevocations,
-		serialBatchSize: config.BadKeyRevoker.FindCertificatesBatchSize,
-		raClient:        rac,
-		mailer:          mailClient,
-		emailSubject:    config.BadKeyRevoker.Mailer.EmailSubject,
-		emailTemplate:   emailTemplate,
-		logger:          logger,
-		clk:             clk,
+		dbMap:               dbMap,
+		maxRevocations:      config.BadKeyRevoker.MaximumRevocations,
+		serialBatchSize:     config.BadKeyRevoker.FindCertificatesBatchSize,
+		raClient:            rac,
+		mailer:              mailClient,
+		emailSubject:        config.BadKeyRevoker.Mailer.EmailSubject,
+		emailTemplate:       emailTemplate,
+		logger:              logger,
+		clk:                 clk,
+		backoffDurationMax:  config.BadKeyRevoker.BackoffDurationMax.Duration,
+		backoffDurationBase: config.BadKeyRevoker.Interval.Duration,
+		backoffFactor:       1.3,
+	}
+
+	// If `BackoffDurationMax` was not set via the config, set it to 60
+	// seconds. This will avoid a tight loop on error but not be an
+	// excessive delay if the config value was not deliberately set.
+	if config.BadKeyRevoker.BackoffDurationMax.Duration == 0 {
+		bkr.backoffDurationMax = time.Second * 60
 	}
+
+	// If `Interval` is not set in the config then set
+	// `bkr.backoffDurationBase` to a default 1 second.
+	if config.BadKeyRevoker.Interval.Duration == 0 {
+		bkr.backoffDurationBase = time.Second
+	}
+
+	// Run bad-key-revoker in a loop. Backoff if no work or errors.
 	for {
 		noWork, err := bkr.invoke()
 		if err != nil {
 			keysProcessed.WithLabelValues("error").Inc()
 			logger.AuditErrf("failed to process blockedKeys row: %s", err)
+			// Increase backoff duration.
+			bkr.backoffTick()
+			logger.Infof("backoff on error, trying again in %f seconds.", bkr.backoffDuration.Seconds())
+			// Sleep for backoff duration.
+			bkr.clk.Sleep(bkr.backoffDuration)
 			continue
 		}
 		if noWork {
-			logger.Info(fmt.Sprintf(
-				"No work to do. Sleeping for %s", config.BadKeyRevoker.Interval.Duration))
-			time.Sleep(config.BadKeyRevoker.Interval.Duration)
+			// Increase backoff duration.
+			bkr.backoffTick()
+			logger.Infof("no work to do, trying again in %f seconds", bkr.backoffDuration.Seconds())
+			// Sleep for `backoffDuration`
+			bkr.clk.Sleep(bkr.backoffDuration)
 		} else {
 			keysProcessed.WithLabelValues("success").Inc()
+			// Successfully processed, reset backoff.
+			bkr.backoffReset()
 		}
 	}
 }
+
+// backoffTick increments the backoffTicker and calls core.RetryBackoff
+// to calculate a new backoff duration and set `bkr.backoffDuration`.
+func (bkr *badKeyRevoker) backoffTick() {
+	bkr.backoffTicker++
+	bkr.backoffDuration = core.RetryBackoff(
+		bkr.backoffTicker,
+		bkr.backoffDurationBase,
+		bkr.backoffDurationMax,
+		bkr.backoffFactor,
+	)
+}
+
+// reset sets the backoff ticker and duration to zero.
+func (bkr *badKeyRevoker) backoffReset() {
+	bkr.backoffTicker = 0
+	bkr.backoffDuration = 0
```

Comment:
Shouldn't this be `bkr.backoffDuration = brk.backoffDurationBase`?

Reply:
bkr.backoffDuration is removed so no longer reset.